### ISSUE TITLE
Potential dependency conflicts between aucome and docopt

### ISSUE THREAD

**Issue by NeolithEra** [AuReMe/aucome#1]
Hi, as shown in the following full dependency graph of  **aucome_**, aucome requires _**docopt**_ (the latest version)， while the installed version of **_padmet_**(3.2.1) requires _**docopt==0.6.2**_.

According to Pip's “first found wins” installation strategy, _**docopt 0.6.2**_ is the actually installed version.  

Although the first found package version _**docopt 0.6.2**_ just satisfies the later dependency constraint （_**docopt==0.6.2**_), it will lead to a build failure once developers release a newer version of docopt.

### Dependency tree------
```
aucome<version range:>
| +-docopt<version range:>
| +-eventlet<version range:>
| +-mpwt<version range:>
| | +-biopython<version range:>=1.70>
| | +-docopt<version range:>=0.6.2>
| | +-gffutils<version range:>=0.9>
| | | +-argcomplete<version range:>
| | | +-argh<version range:>
| | | +-pyfaidx<version range:>
| | | | +-setuptools<version range:>=0.7>
| | | | +-six<version range:>
| | | +-simplejson<version range:>
| | | +-six<version range:>
| +-padmet<version range:>
| | +-biopython<version range:==1.70>
| | +-cobra<version range:==0.14.0>
| | +-docopt<version range:==0.6.2>
| | +-python-libsbml<version range:==5.16.0>
| +-requests<version range:>
```

Thanks for your attention.
Best,
Neolith

**Comment by NeolithEra** [comment]
### Suggestion
1.	Fix your direct dependencies to be **_docopt==0.6.2_** and **_padmet==3.2.1_**, to remove this conflict. 
        I have checked this revision will not affect your downstream projects now.

2.	Ask your upstream project **_padmet_** to loose the version range of **__docopt_** to be **_>=0.6.2_**.

Hi, @ArnaudBelcour Which solution do you prefer, 1 or 2?
Please let me know your choice. I can submit a PR to solve this issue.

**Comment by ArnaudBelcour** [comment]
Hi,

Thanks for the issue.

As you suggested we can fix this by freezing the dependencies.
This is in my TODO list, I am going to freeze all the dependencies in the next weeks.

Thanks again,
Best regards,
Arnaud Belcour.

**Comment by NeolithEra** [comment]
@ArnaudBelcour Thank you very much!
BTW, dependency conflict problems are very common in practice in python ecosystem.
I have encountered the issues several times. What's your oppinion on the challenges of detecting and repairing such issues?
And how to avoid dependency conflict problems among python projects, especially for the overall python ecosystem?

**Comment by ArnaudBelcour** [comment]
This is a very tough question. Especially if you look at the overall python ecosystem. Because in this ecosystem, you will encounter a lot of projects with different goals in mind. For example, you can encounter library that will be imported by other projects and complete workflow. These two types of project will handle differently the dependencies ([for example if we have or not to pin the repository]( https://stackoverflow.com/a/44938662)).

I don’t have a solution to these conflicts other than checking conflict manually or with tools like [pipdeptree]( https://github.com/naiquevin/pipdeptree) or use other tools to install package like [pipx](https://github.com/pipxproject/pipx), which seems to better handle these conflicts than pip.

I also  have a question as I see your activity are you a human researching all these repositories or an automaton/IA made by Microsoft to detect these errors?

Because if you are a human, you have all my respect and gratitude for what you are doing (and if you are an automaton/IA, well you are build for that so sorry it is less impressive)!

**Comment by NeolithEra** [comment]
@ArnaudBelcour Thanks for your feedback.
I am a human (more precisely, a girl).   I just use AI algorithm to do this.
I am not made by Microsoft. 
Build a good dependency ecosystem for python projects is our common goal ^_^.

**Comment by ArnaudBelcour** [comment]
The new release of padmet go further into solving this issue as it allows docopt to be >=0.6.2.

Fix in the AuReMe/padmet#8.
